auto_dev/services/runner/runner.py

```diff
@@ -0,0 +1,301 @@
+
+    def setup_agent(self) -> None:
+        """Setup the agent."""
+        if not self.fetch:
+            self.logger.info(f"Agent author: {self.agent_name.author}")
+            self.logger.info(f"Agent name: {self.agent_name.name}")
+
+        self.logger.info("Setting up agent keys...")
+        self.manage_keys()
+
+        self.logger.info("Installing dependencies...")
+        self.install_dependencies()
+
+        self.logger.info("Setting up certificates...")
+        self.issue_certificates()
+        self.logger.info("Agent setup complete. 🎉")
+
+    def manage_keys(
+        self,
+        generate_keys: bool = True,
+    ) -> None:
+        """Manage keys based on the agent's default ledger configuration."""
+        config = load_autonolas_yaml(PackageType.AGENT)[0]
+        required_ledgers = config["required_ledgers"]
+        if not required_ledgers:
+            self.logger.error("No ledgers found in the agent configuration.")
+            sys.exit(1)
+        for ledger in required_ledgers:
+            self.logger.info(f"Processing ledger: {ledger}")
+            self.setup_ledger_key(ledger, generate_keys)
+
+    def setup_ledger_key(self, ledger: str, generate_keys, existing_key_file: Path | None = None) -> None:
+        """Setup the agent with the ledger key."""
+        key_file = Path(f"{ledger}_private_key.txt")
+        commands_to_errors = []
+        if existing_key_file:
+            self.logger.info(f"Copying existing key file {existing_key_file} to {key_file}")
+            shutil.copy(existing_key_file, key_file)
+        if key_file.exists():
+            self.logger.error(f"Key file {key_file} already exists.")
+        else:
+            if generate_keys:
+                self.logger.info(f"Generating key for {ledger}...")
+                commands_to_errors.append([f"aea -s generate-key {ledger}", f"Key generation failed for {ledger}"])
+            commands_to_errors.append([f"aea -s add-key {ledger}", f"Key addition failed for {ledger}"])
+
+        for command, error in commands_to_errors:
+            result = self.execute_command(command)
+            if not result:
+                self.logger.error(error)
+        self.logger.info(f"{ledger} key setup complete ✅")
+
+    def install_dependencies(self) -> None:
+        """Install agent dependencies."""
+        self.execute_command("aea -s install")
+
+    def issue_certificates(self) -> None:
+        """Issue certificates for agent if needed."""
+        if not Path("../certs").exists():
+            self.execute_command("aea -s issue-certificates")
+        else:
+            self.execute_command("cp -r ../certs ./")
+
+    def start_tendermint(self, env_vars=None) -> None:
+        """Start Tendermint."""
+        self.logger.info("Starting Tendermint with docker-compose...")
+        try:
+            result = self.execute_command(
+                f"docker compose -f {DOCKERCOMPOSE_TEMPLATE_FOLDER}/tendermint.yaml up -d --force-recreate",
```

Comment:
i do think some kind of port conflict logic here would be very powerful at some point

Reply:
not needed now